services/templeton/src/templeton/main.go

```diff
@@ -0,0 +1,177 @@
+	cfgPath            = flag.String("config", "", "Path to the configuration file. Required except for --version.")
+	amqpURI            string
+	elasticsearchBase  string
+	elasticsearchIndex string
+	dbURI              string
+	gitref             string
+	appver             string
+	builtby            string
+)
+
+func init() {
+	flag.Parse()
+}
+
+// AppVersion prints version information to stdout
+func AppVersion() {
+	if appver != "" {
+		fmt.Printf("App-Version: %s\n", appver)
+	}
+	if gitref != "" {
+		fmt.Printf("Git-Ref: %s\n", gitref)
+	}
+
+	if builtby != "" {
+		fmt.Printf("Built-By: %s\n", builtby)
+	}
+}
+
+func checkMode() {
+	validModes := []string{"periodic", "incremental", "full"}
+	foundMode := false
+
+	for _, v := range validModes {
+		if v == *mode {
+			foundMode = true
+		}
+	}
+
+	if !foundMode {
+		fmt.Printf("Invalid mode: %s\n", *mode)
+		flag.PrintDefaults()
+		os.Exit(-1)
+	}
+}
+
+func initConfig(cfgPath string) {
+	err := configurate.Init(cfgPath)
+	if err != nil {
+		logger.Fatal(err)
+	}
+}
+
+func loadElasticsearchConfig() {
+	var err error
+	elasticsearchBase, err = configurate.C.String("elasticsearch.base")
+	if err != nil {
+		logger.Fatal(err)
+	}
+	elasticsearchIndex, err = configurate.C.String("elasticsearch.index")
+	if err != nil {
+		logger.Fatal(err)
+	}
+}
+
+func loadAMQPConfig() {
+	var err error
+	amqpURI, err = configurate.C.String("amqp.uri")
+	if err != nil {
+		logger.Fatal(err)
+	}
+}
+
+func loadDBConfig() {
+	var err error
+	dbURI, err = configurate.C.String("db.uri")
+	if err != nil {
+		logger.Fatal(err)
+	}
+}
+
+func main() {
+	if *version {
+		AppVersion()
+		os.Exit(0)
+	}
+
+	checkMode()
+
+	if *cfgPath == "" {
+		fmt.Println("--config is required")
+		flag.PrintDefaults()
+		os.Exit(-1)
+	}
+
+	initConfig(*cfgPath)
+	loadElasticsearchConfig()
+	es, err := elasticsearch.NewElasticer(elasticsearchBase, elasticsearchIndex)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	defer es.Close()
+
+	loadDBConfig()
+	d, err := database.NewDatabaser(dbURI)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	if *mode == "full" {
+		logger.Println("Full indexing mode selected.")
+
+		es.Reindex(d)
+		return
+	}
+
+	loadAMQPConfig()
+
+	client, err := messaging.NewClient(amqpURI, true)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	defer client.Close()
+
+	if *mode == "periodic" {
+		logger.Println("Periodic indexing mode selected.")
+
+		go client.Listen()
+
+		// Accept and handle messages sent out with the index.all and index.templates routing keys
+		client.AddConsumer(messaging.ReindexExchange, "direct", "templeton.reindexAll", messaging.ReindexAllKey, func(del amqp.Delivery) {
+			es.Reindex(d)
+			del.Ack(false)
+		})
+		client.AddConsumer(messaging.ReindexExchange, "direct", "templeton.reindexTemplates", messaging.ReindexTemplatesKey, func(del amqp.Delivery) {
+			es.Reindex(d)
+			del.Ack(false)
+		})
```

Comment:
A modification/improvement to this would presumably be to listen to both types of messages on one queue. I don't think this is required initially, but I could be convinced.